docs/spec/supervisor/supervisor.md

```diff
@@ -0,0 +1,153 @@
+/// highest one for which we believe originated from the same chain
+type ChainPair = (LightBlock, Height);
+
+/// Traces the history of verify_to_target.
+trait Trace {
+    /// Returns the height of the current highest inserted
+    fn current_height(&self) -> Height;
+
+    /// Returns the current highest inserted light block.
+    fn current(&self) -> &LightBlock;
+
+    /// Insert a new highest light block. assert(self.current_height() < light_block.height())
+    fn insert(&mut self, light_block: LightBlock) -> Result<(), Error>;
+
+    /// Returns the list of all LightBlock which were inserted through 
+    /// insert with associated previous current_height().
+    fn verification_chain(&self) -> LinkedList<ChainPair>;
+}
+
+
+/// Given a primary and a starting_light_block confirm that the light block of target_height received
+/// from primary for target_height is from the same chain as the starting_light_block.
+/// Solution for forward skipping verification.
+pub fn verify_to_target(
+    primary: PeerID,
+    starting_light_block: LightBlock,
+    target_height: Height) -> Result<Trace, (LightBlock, Reason)> {
+
+    // Verified state always has one starting LightBlock
+    let mut trace = Trace::from(startingLightBlock);
+    let mut fetch_stack = FetchedStack::empty();
+
+    // insert the light block of target height, simplifies the loop
+    // we can say that at each iteration there should be something on stack
+    fetch_stack.push(fetch_light_block(primary, target_height));
+
+    while trace.current_height() < targetHeight {
+        // unverified state is never empty
+        assert!(!fetch_stack.is_empty());
+        // always try with previously received
+        let lowest_unverified = fetch_stack.peek().unwrap().clone();
+        
+        // do the necessary checks
+        let verdict = valid_and_verified(trace.current(), &lowest_unverified);
+
+        match verdict {
+            // validated LightBlock, improve the verified_state
+            Valid => {
+                fetch_stack.pop();
+                trace.insert(lowest_unverified);
+                continue;
+            },
+            // invalid block stop
+            Invalid(reason) => {
+                return Err((lowest_unverified, reason))
+            },
+            Untrusted => {
+                // we need an intermediate LightBlock
+                current_height = 
+                    compute_height(trace.current_height(), lowest_unverified.height());
+            }
+        }
+    }
+
+    assert_eq!(trace.current_height(), target_height);
+    assert!(fetch_stack.is_empty());
+
+    return Ok(trace);
+}
+
+/// LightStore no longer has notions of Trusted, Verified, Unverified and Failed.
+/// The LightBlocks which are stored are all assumed to have originated from the
+/// same chain, irrespective of the peer from which we received it. The store
+/// provides the invariant that there is a 1-1 mapping from height to a LightBlock.
```

Comment:
If we want to ensure the 1-1 invariant we should think in terms of a `Set` where we have a uniqueness constraint on the height, unless I miss-understand this part.

Reply:
I think it would make more sense to have a `Map<Height, LightBlock>` where keys would be automatically unique.

Reply:
For clarification, every height can only exist once? And there is only exactly one `LightBlock` per unique height?

Reply:
Yes, that was the intention. At each point in time, there can be only one `LightBlock` which can be considered to be from the valid blockchain. Conflict on a single height would mean there is a fork.